### PR TITLE
Check for any segments not in preferred roles

### PR DESCRIPTION
**Pull request by bhuvnesh2703**
pg_upgrade check should identify (fail) if there are any segments
not in their preferred roles

The below error will be seen
```
Checking for segments not in preferred roles                fatal

| The segments should be in their preferred roles for upgrade.
| Segments not operating in the preferred roles must be recovered
| and/or rebalanced using gprecoverseg, the list of such segments
| is in the file:
	segment_roles_switched.txt

Failure, exiting
✘-1 ~/workspace/gpdb [add-role-check-upgrade L|…51⚑ 1]
12:38 $ cat segment_roles_switched.txt
dbid "3" with content id "1" has switched role: role: "m" preferred_role: "p"
dbid "6" with content id "1" has switched role: role: "p" preferred_role: "m"
✔ ~/workspace/gpdb [add-role-check-upgrade L|…51⚑ 1]
12:38 $
```

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: contrib/pg_upgrade/greenplum/check_gp.c [hunk 1]
```diff
@@ -897,3 +899,72 @@ check_for_array_of_partition_table_types(ClusterInfo *cluster)
 
 	check_ok();
 }
+
+/*
+ *	check_preferred_roles
+ *
+ *	Check for any segment which is not currently in their preferred roles
```
Comment (dkrieger-pvtl) [inline contrib/pg_upgrade/greenplum/check_gp.c]: nit: `Check for any segment which is not currently in its preferred role` (keep consistent singular usage)

File: contrib/pg_upgrade/greenplum/check_gp.c [hunk 2]
```diff
@@ -897,3 +899,72 @@ check_for_array_of_partition_table_types(ClusterInfo *cluster)
...
+	PQfinish(conn);
+
+	if (ntups > 0)
+	{
+		fclose(script);
+		pg_log(PG_REPORT, "fatal\n");
+		pg_log(PG_FATAL,
+			   "| The segments should be in their preferred roles for upgrade.\n"
+			   "| Segments not operating in the preferred roles must be recovered \n"
+			   "| and/or rebalanced using gprecoverseg, the list of such segments \n"
```
Comment (dkrieger-pvtl) [inline contrib/pg_upgrade/greenplum/check_gp.c]: nit: should read `"| and/or rebalanced using gprecoverseg. The list of such segments \n"

File: contrib/pg_upgrade/greenplum/check_gp.c [hunk 3]
```diff
@@ -897,3 +899,72 @@ check_for_array_of_partition_table_types(ClusterInfo *cluster)
...
+			   "| is in the file:\n\t%s\n\n", output_path);
+	}
+	else
+		check_ok();
+}
```
Comment (dkrieger-pvtl) [inline contrib/pg_upgrade/greenplum/check_gp.c]: nit: place a new-line here.

**Comment by bhuvnesh2703** [comment]
applied feedback and added the tests..

File: contrib/pg_upgrade/test/integration/input/preferred_roles.source [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+checkpoint;
+-- the check should not return any failure
+! ./gpdb6/bin/pg_upgrade --mode=dispatcher --old-gp-dbid=1 --new-gp-dbid=1 --check --old-bindir=@upgrade_test_path@/gpdb5/bin --new-bindir=@upgrade_test_path@/gpdb6/bin --old-datadir=@upgrade_test_path@/gpdb5-data/qddir/demoDataDir-1 --new-datadir=@upgrade_test_path@/gpdb6-data/qddir/demoDataDir-1 --old-port=@old_port@;
+
+-- Stop content 0 primary and let the mirror take over
+select stop_segment(fselocation) from pg_filespace_entry fe, gp_segment_configuration c, pg_filespace f
+where fe.fsedbid = c.dbid and c.content=0 and c.role='p' and f.oid = fe.fsefsoid and f.fsname = 'pg_system';
+
+select wait_for_content0('c');
```
Comment (dkrieger-pvtl) [inline contrib/pg_upgrade/test/integration/input/preferred_roles.source]: Is this the correct state change to wait for?  I thought you'd check the `content 0` has a promoted primary instead(the same check done in this PR for `pg_upgrade`).

**Comment by dkrieger-pvtl** [comment]
Is `pg_upgrade` the right place to put this preferred roles check(as opposed to [gpupgrade](https://github.com/greenplum-db/gpupgrade))?  The check you do here is a cluster-wide check, and `pg_upgrade` is a single-postgres-instance check.  Conceptually, I think it belongs in pg_upgrade.  Even for a mismatched role for a single instance ( [pg_upgrade doc](https://www.postgresql.org/docs/9.4/pgupgrade.html)), there seems to be nothing wrong with upgrading a promoted instance(itself a Greenplum concept).

**Comment by jchampio** [comment]
> Conceptually, I think it belongs in pg_upgrade.

@dkrieger-pvtl Is this what you meant? The rest of your post seems to be arguing for `gpupgrade`, but maybe I misunderstood?

If we put this check in `pg_upgrade`, the workflow looks like this if a user has an unbalanced cluster:
- `gpupgrade initialize` pulls the unbalanced configuration,
- creates a new cluster based on the wrong information, and
- fails in `pg_upgrade --check`.
- Then we make them `gpupgrade revert` and start over.

Seems like we should move that check earlier, before the new cluster is ever created to begin with.

**Comment by dkrieger-pvtl** [comment]
> > Conceptually, I think it belongs in pg_upgrade.
> 
> @dkrieger-pvtl Is this what you meant? The rest of your post seems to be arguing for `gpupgrade`, but maybe I misunderstood?


Sorry....I mis-wrote.  I meant I think this check should be in `gpupgrade` entirely.

**Comment by bhuvnesh2703** [comment]
Closing this PR, introduced this check in gpugprade. https://github.com/greenplum-db/gpupgrade/pull/219/files

**Pull request closed without merging**